File: 08/winter/family_tree.py
line = raw_input()
everyone = set()
descendants = {}
ancestors = {}
while line:
    father, mother, child = line.split()
    descendants.setdefault(father, set()).add(child)
    descendants.setdefault(mother, set()).add(child)
    ancestors.setdefault(child, set()).add(mother)
    ancestors[child].add(father)
    everyone.update(set([father, mother, child]))
    try: line = raw_input()
    except: break

for parent in everyone:
    oldsize = len(descendants.get(parent, []))
    for kid in list(descendants.get(parent, []))[:]:
        descendants[parent].update(descendants.get(kid, set()))
    newsize = len(descendants.get(parent, []))
    while newsize > oldsize:
        oldsize = len(descendants.get(parent, []))
        for kid in list(descendants.get(parent, []))[:]:
            descendants[parent].update(descendants.get(kid, set()))
        newsize = len(descendants.get(parent, []))

for kid in everyone:
    oldsize = len(ancestors.get(kid, []))
    for parent in list(ancestors.get(kid, []))[:]:
        ancestors[kid].update(ancestors.get(parent, set()))
    newsize = len(ancestors.get(kid, []))
    while newsize > oldsize:
        oldsize = len(ancestors.get(kid, []))
        for parent in list(ancestors.get(kid, []))[:]:
            ancestors[kid].update(ancestors.get(parent, set()))
        newsize = len(ancestors.get(kid, []))
        
#print descendants
#print ancestors
for x in raw_input().split():
    print x, "has", len(descendants.get(x, [])), "descendants and", \
          len(ancestors.get(x, [])), "ancestors"

File: 08/winter/solutions/family_tree.java
import java.util.*;

class family_tree {

	private static final Map<String,List<String>> children = new HashMap<String,List<String>>();
	private static final Map<String,List<String>> parents = new HashMap<String,List<String>>();

	private static Set<String> descendants(String person) {

		Set<String> descendants = new HashSet<String>();
		List<String> l = children.get(person);
		if (l != null) {
			descendants.addAll(l);
			for (String child : l)
				descendants.addAll(descendants(child));
		}
		return descendants;
	}

	private static Set<String> ancestors(String person) {

		Set<String> ancestors = new HashSet<String>();
		List<String> l = parents.get(person);
		if (l != null) {
			ancestors.addAll(l);
			for (String parent : l)
				ancestors.addAll(ancestors(parent));
		}
		return ancestors;
	}

	public static void main(String[] arguments) {

		Scanner in = new Scanner(System.in);

		while (in.hasNextLine()) {

			Scanner line = new Scanner(in.nextLine());

			if (! line.hasNext()) break;

			String father = line.next();
			String mother = line.next();
			String child = line.next();

			List<String> l;

			l = children.get(father);
			if (l == null) l = new LinkedList<String>();
			l.add(child);
			children.put(father, l);

			l = children.get(mother);
			if (l == null) l = new LinkedList<String>();
			l.add(child);
			children.put(mother, l);

			l = new LinkedList<String>();
			l.add(father);
			l.add(mother);
			parents.put(child, l);
		}
		while (in.hasNext()) {

			String person = in.next();
			System.out.printf("%s has %d descendants and %d ancestors%n",
				person, descendants(person).size(), ancestors(person).size());
		}
	}
}
